zip-0312.rst

```diff
@@ -1,7 +1,466 @@
+- Like the FROST specification [#FROST], this ZIP does not specify a key generation
+  procedure; but refer to that specification for guidelines.
+- Network privacy is not in scope for this ZIP, and must be obtained with other
+  tools if desired.
+
+
+Specification
+=============
+
+Algorithms in this section are specified using Python pseudo-code, in the same
+fashion as the FROST specification [#FROST]_.
+
+The types Scalar, Element, and G are defined in #[frost-primeordergroup]_, as well
+as the notation for elliptic-curve arithmetic, which uses the additive notation.
+
+
+Re-randomizable FROST
+---------------------
+
+To add re-randomization to FROST, follow the specification [#FROST]_ with the
+following modifications.
+
+
+Key Generation
+''''''''''''''
+
+While key generation is out of scope for this ZIP and the FROST spec [#FROST]_,
+it needs to be consistent with FROST, see [#frost-tdkg]_ for guidance. The
+spend authorization private key :math:`\mathsf{ask}` [#protocol-spendauthsig]_
+is the particular key that must be used in the context of this ZIP. Note that
+while Sapling allows creating it directly, in Orchard it is always derived
+from the spending key :math:`\mathsf{ask}`. This means that a trusted
+dealer key generation process might be required as detailed in [#frost-tdkg]_.
+
+
+Randomizer Generation
+'''''''''''''''''''''
+
+A new helper function is defined, which computes :math:`\mathsf{RedDSA.GenRandom}`:
+
+::
+
+  randomizer_generate():
+
+  Inputs:
+  - None
+
+  Outputs: randomizer, a Scalar
+
+  def randomizer_generate():
+    randomizer_input = random_bytes(64)
+    return H3(randomizer_input)
+
+
+Binding Factor Computation
+''''''''''''''''''''''''''
+
+The `compute_binding_factor` function is changed to receive the `randomizer_point`
+as follows: ::
+
+  Inputs:
+  - commitment_list = [(i, hiding_nonce_commitment_i, binding_nonce_commitment_i), ...],
+    a list of commitments issued by each participant, where each element in the list
+    indicates a NonZeroScalar identifier i and two commitment Element values
+    (hiding_nonce_commitment_i, binding_nonce_commitment_i). This list MUST be sorted
+    in ascending order by identifier.
+  - msg, the message to be signed.
+  - randomizer_point, an element in G.
+
+  Outputs:
+  - binding_factor_list, a list of (NonZeroScalar, Scalar) tuples representing the binding factors.
+
+  def compute_binding_factors(commitment_list, msg, randomizer_point):
+    msg_hash = H4(msg)
+    encoded_commitment_hash = H5(encode_group_commitment_list(commitment_list))
+    rho_input_prefix = msg_hash || encoded_commitment_hash
+
+    binding_factor_list = []
+    for (identifier, hiding_nonce_commitment, binding_nonce_commitment) in commitment_list:
+      rho_input = rho_input_prefix || G.SerializeScalar(identifier)
+      binding_factor = H1(rho_input)
+      binding_factor_list.append((identifier, binding_factor))
+    return binding_factor_list
```

Comment:
This does not use `randomizer_point`. Was it intended to pass an encoding of it to H5 or include it in `rho_input_prefix`?